backend/cfg/cfg_dataflow.ml

```diff
@@ -193,88 +414,86 @@ module type Backward_S = sig
     ('a, error) Dataflow_result.t
 end
 
-module Backward
-    (D : Backward_domain)
-    (T : Backward_transfer with type domain = D.t) :
+module Backward (D : Domain_S) (T : Backward_transfer with type domain = D.t) :
   Backward_S with type domain = D.t and type error = T.error = struct
-  (* CR xclerc for xclerc: see what can be shared with `Forward`. *)
-
-  type domain = D.t
-
   type error = T.error
 
   exception Dataflow_aborted of error
 
-  type _ map =
-    | Block : domain Label.Tbl.t map
-    | Instr : domain Instr.Tbl.t map
-    | Both : (domain Instr.Tbl.t * domain Label.Tbl.t) map
-
-  module WorkSetElement = struct
-    type t =
-      { label : Label.t;
-        value : domain
-      }
-
-    let compare { label = left_label; value = left_value }
-        { label = right_label; value = right_value } =
-      match Label.compare left_label right_label with
-      | 0 -> D.compare left_value right_value
-      | res -> res
-  end
-
-  module WorkSet = Set.Make (WorkSetElement)
+  type transfer_domain =
+    { normal : D.t;
+      exn : D.t
+    }
 
   let unwrap_transfer_result value =
     match value with
     | Ok value -> value
     | Error error -> raise (Dataflow_aborted error)
 
-  let transfer_block :
-      domain Instr.Tbl.t option ->
-      domain ->
-      exn:domain ->
-      Cfg.basic_block ->
-      domain =
-   fun tbl value ~exn block ->
-    let replace (instr : _ Cfg.instruction) value =
-      let value = unwrap_transfer_result value in
-      match tbl with
-      | None -> value
-      | Some tbl ->
-        Instr.Tbl.replace tbl instr.id value;
+  module Direction :
+    Dataflow_direction_S
+      with type Transfer_domain.t = transfer_domain
+       and type instr_domain = D.t = struct
+    module Transfer_domain : Transfer_domain_S with type t = transfer_domain =
+    struct
+      type t = transfer_domain =
+        { normal : D.t;
+          exn : D.t
+        }
+
+      let bot = { normal = D.bot; exn = D.bot }
+
+      let less_equal t1 t2 =
+        D.less_equal t1.normal t2.normal && D.less_equal t1.exn t2.exn
+    end
+
+    type transfer_image = D.t
+
+    type instr_domain = D.t
+
+    let edges_out : Cfg.basic_block -> Label.Set.t =
+     fun block -> Cfg.predecessor_labels block |> Label.Set.of_list
```

Comment:
(Possibly technically not part of
this PR, but we may want to avoid
going through the list to build the
set.)

Reply:
we could simply do `val edges_out : Cfg.basic_block -> Label.t Seq.t`
What do you think of that?

Reply:
Changed the return type to `Label.t Seq.t` This is the best common denominator I believe.

Reply:
It is probably good enough, but
you should also not feel constrained
by the current interface of `Cfg`.
If you think an additional accessor
in `Cfg` will make the code in this
module more effecient, that is
definitely an option.

Reply:
I would leave this like this